### PR TITLE
EXERCISE_02

### PR DESCRIPTION
**Pull request by MayDruyan**
This is the first time I am doing a pull request. Very cool! This is also the first time I am writing an Android app :)

### CODE REVIEW

File: firstapp/app/src/main/java/com/example/mygameofthronesapp/Data.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.example.mygameofthronesapp;
+
+public class Data {
```
Comment (talboren) [inline firstapp/app/src/main/java/com/example/mygameofthronesapp/Data.java]: Classes best practice is to have intuitive names that can infer the usage of the class.
A better name could have been "Character" or "Person"

File: firstapp/app/src/main/java/com/example/mygameofthronesapp/Data.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    private String name;
+    private int image;
+
+    public Data(String name, int image) {
+        this.name = name;
+        this.image = image;
+    }
+
+    public String getName() {
```
Comment (talboren) [inline firstapp/app/src/main/java/com/example/mygameofthronesapp/Data.java]: Great job using private objects and exposing them with a getter!

File: firstapp/app/src/main/java/com/example/mygameofthronesapp/LannisterFamilyActivity.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            R.drawable.tyrionlannister,R.drawable.myrcellalannister,R.drawable.tommenlannister};
+    private int PORTRAIT_IMAGE = R.drawable.lannister_port;
+    private int LANDSCAPE_IMAGE = R.drawable.lannister_land;
+
+    /**
+     * This method initializes the data that needs to be on this activity
+     */
+    private void createData() {
+        this.myData = new ArrayList<>();
+        for (int i = 0; i < charNames.length; i++) {
```
Comment (talboren) [inline firstapp/app/src/main/java/com/example/mygameofthronesapp/LannisterFamilyActivity.java]: Why not creating a list with the "Data" objects already initiated?
See the example in the comment above

File: firstapp/app/src/main/java/com/example/mygameofthronesapp/LannisterFamilyActivity.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import android.util.Log;
+import java.util.ArrayList;
+
+import androidx.appcompat.app.AppCompatActivity;
+import androidx.recyclerview.widget.LinearLayoutManager;
+import androidx.recyclerview.widget.RecyclerView;
+
+public class LannisterFamilyActivity extends AppCompatActivity {
+    private static final String TAG = "LannisterActivity";
+    private ArrayList<Data> myData;
```
Comment (talboren) [inline firstapp/app/src/main/java/com/example/mygameofthronesapp/LannisterFamilyActivity.java]: = new ArrayList<Data>() { new Data("Tywin", R.drawable.tywinlannister ... }

File: firstapp/app/src/main/java/com/example/mygameofthronesapp/StarkFamilyActivity.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    private String[] charNames = {"Eddard","Catelyn","Robb","Arya","Sansa","Bran","Rickon"};
+    private int[] charImages = {R.drawable.eddardstark,R.drawable.catelynstark,R.drawable.robbstark,
+            R.drawable.aryastark,R.drawable.sansastark,R.drawable.branstark, R.drawable.rickonstark};
+    private int PORTRAIT_IMAGE = R.drawable.stark_port;
+    private int LANDSCAPE_IMAGE = R.drawable.stark_land;
+
+    /**
+     * This method initializes the data that needs to be on this activity
+     */
+    private void createData() {
```
Comment (talboren) [inline firstapp/app/src/main/java/com/example/mygameofthronesapp/StarkFamilyActivity.java]: See the comments from the LannisterFamilyActivitty.
The data initialization process here is redundant.

File: MyNotificationsApp/app/src/main/java/com/example/mynotificationsapp/AlarmReceiver.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import android.util.Log;
+import androidx.core.app.NotificationCompat;
+
+/**
+ * This class is triggered by AlarmService. Sends notification with quote
+ */
+public class AlarmReceiver extends BroadcastReceiver {
+    private static final String TAG = "AlarmReceiver";
+    private static final String CHANNEL_ID = "NotificationAppChannel";
+    private final String notificationTitle = "Your quote!";
```
Comment (talboren) [inline MyNotificationsApp/app/src/main/java/com/example/mynotificationsapp/AlarmReceiver.java]: Great!

File: MyNotificationsApp/app/src/main/java/com/example/mynotificationsapp/AlarmReceiver.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, CHANNEL_ID)
+                .setSmallIcon(R.drawable.ic_filter_vintage_pink)
+                .setContentTitle(notificationTitle)
+                .setStyle(new NotificationCompat.BigTextStyle().bigText(curQuote))
+                .setContentIntent(pendingIntent)
+                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
+                .setAutoCancel(true);
+
+        nm.notify(notificationId, builder.build());
+        Log.d(TAG, "Notification is sent with notification id " + notificationId);
```
Comment (talboren) [inline MyNotificationsApp/app/src/main/java/com/example/mynotificationsapp/AlarmReceiver.java]: Love the usage of logs :)

File: MyNotificationsApp/app/src/main/java/com/example/mynotificationsapp/QuotesList.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        String tmp = myData.get(ptr);
+        myData.set(ptr,myData.get(j));
+        myData.set(j, tmp);
+        if (ptr > 0) {
+            ptr--;
+        } else {
+            ptr = arraySize - 1;
+        }
+        return myCurrentQuote;
+    }
```
Comment (talboren) [inline MyNotificationsApp/app/src/main/java/com/example/mynotificationsapp/QuotesList.java]: This function feels a little bit "over-headed" - it is pretty difficult to understand and it's outcome could be done a lot more easier.

File: HttpServerInNodeJS/index.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        let newTaskId = req.query.id;
+        let description = req.query.task;
+        
+        // Add the new task
+        data[newTaskId] = description;
+        
+        // Rewrite the data to the file
+        fs.writeFile(dataPath, JSON.stringify(data, null, 2), (err) => {
+            if (err) throw err;                
+            res.status(200).send(`New task with id : ${newTaskId} and description : "${description}" added!`);
```
Comment (ransh93) [inline HttpServerInNodeJS/index.js]: Good usage of formatting

File: HttpServerInNodeJS/index.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+const PORT = 8000;
+
+let app = express();
+app.use(bodyParser.json());
+
+// Read tasks list
+app.get('/tasks', (req, res) => {
+    console.log('Got GET request to /tasks')
+    fs.readFile(dataPath, (err, data) => {
+        if (err) throw err;
```
Comment (ransh93) [inline HttpServerInNodeJS/index.js]: It is better to return a clear error to the user so he will know there was a problem along the way

File: HttpServerInNodeJS/index.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        res.send(JSON.parse(data));
+    });
+});
+
+// Create a new task
+app.post('/tasks/new', (req, res) => {
+    console.log('Got POST request to /tasks/new');
+    fs.readFile(dataPath, (err, data)=> {
+        if (err) throw err;
```
Comment (ransh93) [inline HttpServerInNodeJS/index.js]: Same

**Pull request merged**